STIPS/STIP-002.md

```diff
@@ -0,0 +1,150 @@
+
+## Feasibility Analysis
+
+### Option 1:
+Instead of over-optimizing our checks to test for both undercollaterlization and overcollaterlization. We can just have checks to prevent undercollaterlization. This would prevent reverting in cases when `newBalance` of aToken is greater than `existingBalance + quantity` by 1 wei, and thus decrease the failure rate of issuance/redemption (maybe to 33%, cause we can consider less than, equal to, and greater than to be 3 equal cases). This option is the best considering time. The change is contained within the smart contracts and doesn't bubble up the stack leading to bigger changes in our backend.
+
+
+
+## Open Questions
+
+1. Do we restrict usage of these modified issuance modules (which are linked to the new libraries) from third-party use?
+    - Why did we have the strict checks in the first place?
+        - To make sure the set is not undercollateralized 
+            - Eg. To stop tokens which collects fees on transfer from being a component in a SetToken
+        - This strict check achieved the above requirement, but at the same time it also prevented overcollateralization
+    - What are the only requirements for our system to function well?
+        - We do not want Set to be undercollateralized
+        - We are fine with overcollaterlization (discussed below)
+    - Thus, the new checks full fill our requirements.
+        - So, we need not restrict usage of these modified issuance modules
+
+2. Do we restrict these new modified issuance modules (which are linked to the new libraries) only to SetToken which hold aTokens? Or do we make these issuance modules the default going forward?
+    
+2. Are we fine with Set being overcollaterlized?
+    - Yes, SetToken can resync the extra assets present to increase their default positions with help of modules
+        - ALM and CLM can resync positions for leveraged tokens
+        - Airdrop module can add extra assets as positions for other SetTokens
+    - Also, issuance/redemption generally would not lead to overcollaterlization
+        - For it to happen, we would need the special case of a token which rewards holders on a transfer
+        - we can handle that case using Airdrop module
+
+3. How to handle `newBalance < existingBalance.add(_quantity))`?
+    - This is a possible case for aTokens, since `(quantity/index) * index` can be less than `quantity`.
+    - We have seen this case once during testing.
+
+
+## Checkpoint 1
+
+**Reviewer**:
+
+## Proposed Architecture Changes
+
+Pseudocode of option1 :
+```javascript
+
+library Invoke {
+    ... // other functions from Invoke library
+    function strictInvokeTransfer(ISetToken _setToken, address _token, address _to, uint256 _quantity) internal {
+        ../// existing implementations
+    }
+    function invokeTransferWithCollaterlizationChecks(ISetToken _setToken, address _token, address _to, uint256 _quantity) internal {
+        if (_quantity > 0) {
```

Comment:
I don't think this would be a meaningful improvement over the current status quo unless there was some proof that aToken balances are biased towards rounding up. This seems like it would still revert 33% of the time. What I would propose is the following check:

```
            // Get position unit of transferred token for SetToken
            // Get new total supply
            Invoke.invokeTransfer(_setToken, _token, _to, _quantity);
            // Get new balance of transferred token for SetToken
            uint256 newBalance = IERC20(_token).balanceOf(address(_setToken));
            // prevent undercollateralization
            require(newBalance >= newTotalSupply.preciseMul(positionUnit)), "Invalid post transfer balance");
 ```
 This may require changing the interface a bit. This is less stringent than the solution you proposed BUT still guarantees over-collateralization and will lead to less reverts. Should think of if there are any other attack vectors other than a token that charges a fee getting any excess wei in the Set (attack not likely and not very profitable unless positions haven't been synced)